pkg/controllers/vdb/installpackages_reconciler.go

```diff
@@ -0,0 +1,109 @@
+
+import (
+	"context"
+	"time"
+
+	vapi "github.com/vertica/vertica-kubernetes/api/v1"
+	"github.com/vertica/vertica-kubernetes/pkg/cmds"
+	"github.com/vertica/vertica-kubernetes/pkg/controllers"
+	"github.com/vertica/vertica-kubernetes/pkg/events"
+	"github.com/vertica/vertica-kubernetes/pkg/vadmin"
+	"github.com/vertica/vertica-kubernetes/pkg/vadmin/opts/installpackages"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+// InstallPackagesReconciler will install all packages under /opt/vertica/packages where Autoinstall is marked true
+type InstallPackagesReconciler struct {
+	VRec       *VerticaDBReconciler
+	Vdb        *vapi.VerticaDB // Vdb is the CRD we are acting on.
+	PRunner    cmds.PodRunner
+	PFacts     *PodFacts
+	Dispatcher vadmin.Dispatcher
+}
+
+// MakeInstallPackagesReconciler will build a InstallPackagesReconciler object
+func MakeInstallPackagesReconciler(
+	vdbrecon *VerticaDBReconciler, vdb *vapi.VerticaDB, prunner cmds.PodRunner, pfacts *PodFacts,
+	dispatcher vadmin.Dispatcher,
+) controllers.ReconcileActor {
+	return &InstallPackagesReconciler{
+		VRec:       vdbrecon,
+		Vdb:        vdb,
+		PRunner:    prunner,
+		PFacts:     pfacts,
+		Dispatcher: dispatcher,
+	}
+}
+
+// Reconcile will force install default packages in the running database
+func (s *InstallPackagesReconciler) Reconcile(ctx context.Context, _ *ctrl.Request) (ctrl.Result, error) {
+	err := s.PFacts.Collect(ctx, s.Vdb)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// No-op if no database exists
+	if !s.PFacts.doesDBExist() {
+		return ctrl.Result{}, nil
+	}
+
+	// Force reinstall default packages
+	if s.PFacts.getUpNodeCount() > 0 {
+		err = s.installPackagesInPod(ctx)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, err
+}
+
+// installPackagesInPod will find one pod to initiate the process of installing default packages
+func (s *InstallPackagesReconciler) installPackagesInPod(ctx context.Context) error {
+	pf, ok := s.PFacts.findPodToRunAdminCmdAny()
+	if !ok {
+		// If no running pod found, then there is no where to install packages
+		// and we can just continue on
+		return nil
+	}
+
+	// Run the install_packages command
+	err := s.runCmd(ctx, pf.name, pf.podIP)
+
+	return err
+}
+
+// runCmd issues the admintools or vclusterops command to force install the default packages
+func (s *InstallPackagesReconciler) runCmd(ctx context.Context, initiatorName types.NamespacedName, initiatorIP string) error {
+	s.VRec.Event(s.Vdb, corev1.EventTypeNormal, events.InstallPackagesStarted, "Starting install packages")
+	start := time.Now()
+	opts := []installpackages.Option{
+		installpackages.WithInitiator(initiatorName, initiatorIP),
+		installpackages.WithForceReinstall(true),
+	}
+	if err := s.Dispatcher.InstallPackages(ctx, opts...); err != nil {
+		s.VRec.Event(s.Vdb, corev1.EventTypeWarning, events.InstallPackagesFailed, "Failed to install packages")
```

Comment:
If you use `v.logFailure` in `install_packages_vc.go` you don't need this. See my comment below.

Reply:
will do